src/util/expr_util.h

```diff
@@ -72,4 +72,23 @@ bool has_subtype(const typet &, const irep_idt &id, const namespacet &);
 /// lift up an if_exprt one level
 if_exprt lift_if(const exprt &, std::size_t operand_number);
 
+/// Determine whether an expression is constant.  A literal constant is
+/// constant, but so are, e.g., sums over constants or addresses of objects.
+/// An implementation derive from this class to refine what it considers
+/// constant in a particular context by overriding is_constant and/or
+/// is_constant_address_of.
+class is_constantt
```

Comment:
This class doesn't have any state, so I suggest these should be free functions instead. The class is basically acting as a namespace, which the file (with `static` functions) will do anyhow.

Reply:
The "state" is the virtual table - but maybe that's poor design and I should be doing it differently instead?

Reply:
Ah yes I see, sorry, too early to parse code accurately :) I can't think of a better way of doing this, so please disregard the original comment!